### PR TITLE
[5.8] Add `toArray()` to Paginator contract

### PR DESCRIPTION
**Pull request by tomlankhorst**
- Laravel Version: 5.7.13
- PHP Version: 7.2

### Description:
It's possible to paginate eloquent results and [use them in a JSON](https://laravel.com/docs/5.7/pagination#converting-results-to-json) response: 

```php
Route::get('users', function () {
    return App\User::paginate();
});
```

I wanted to send some additional data with the response. I `Paginator` converted it `toArray()` and then added my data:

```php
Route::get('users', function () {
    $result = App\User::paginate()->toArray();
    $result['additional_data'] = // ... 
    return $result;
});
```

As `::paginate()` returns a `Paginator`-contract/-interface, I wondered whether `toArray()` could be added to it? That would ease code-completion and analysis.

### CODE REVIEW

**Comment by Chrisx84** [comment]
if you put the paginate results into a json response it displays as json and it does return as a object in php
$result = App\User::paginate();
$result->additional_data;

**Comment by tomlankhorst** [comment]
That might work for passing around data in the app, however when it's transformed into a request, the `toJson()` -> `jsonSerialize()` -> `toArray()` methods are called... discarding any props on the `Paginator` object.

**Comment by Chrisx84** [comment]
not really when done like this.
`$db = DB::table('sometable')->paginate(25);
return response()->json($db, 200)
->header('Content-Type', 'application/json')
->header('charset', 'utf-8');`

**Comment by tomlankhorst** [comment]
Are you sure? A simple test in Tinker didn't show so.

Op wo 19 dec. 2018 om 20:56 schreef Ven the Meerkat <
notifications@github.com>:

**Comment by Chrisx84** [comment]
ya i know so cuz i do it to all my http requests that doesn't require a html reply. makes things easier for me to parse in the linden scripting language.

**Comment by Chrisx84** [comment]
yup ok it does work
`$db = DB::table('sims')->select('name')->orderBy('name')->simplePaginate(1);
				return response()->json($db, 200) ->header('Content-Type', 'application/json') ->header('charset', 'utf-8');`
which returns
`{"current_page":1,"data":[{"name":"Achemon"}],"first_page_url":"https:\/\/rtcsl.info\/admin\/jtest?page=1","from":1,"next_page_url":"https:\/\/rtcsl.info\/admin\/jtest?page=2","path":"https:\/\/rtcsl.info\/admin\/jtest","per_page":1,"prev_page_url":null,"to":1}`

**Comment by X-Coder264** [comment]
Adding a method to an interface is a breaking change and as such this PR should target the master branch (Laravel 5.8).

And instead of manually adding the `toArray` method to the interface why didn't you make the `Paginator` extend the `Arrayable` interface?

**Comment by tomlankhorst** [comment]
> Adding a method to an interface is a breaking change and as such this PR should target the master branch (Laravel 5.8).
> 
> And instead of manually adding the `toArray` method to the interface why didn't you make the `Paginator` extend the `Arrayable` interface?

`Paginator` does not look like an 'extended Arrayable' to me and single inheritance then limits future additions to the interface imho.

**Comment by X-Coder264** [comment]
The concrete `Paginator` implementation has that method exactly because it implements `Arrayable`. So I don't see any reason why the `Paginator` interface couldn't extend the `Arrayable` interface.

**Comment by tomlankhorst** [comment]
> The concrete `Paginator` implementation has that method exactly because it implements `Arrayable`. So I don't see any reason why the `Paginator` interface couldn't extend the `Arrayable` interface.

That's an interesting observation ;). Seems right. Will update.

**Comment by X-Coder264** [comment]
Now you can also remove "implements `Arrayable`" from the concrete `Paginator` implementation since that is covered now already by the `Paginator` contract :)

**Comment by tomlankhorst** [comment]
I see... however I've reverted back to adding the 'toArray' because it's arbitrary which `Illuminate/Contracts/Support/` extends as it implements both `Jsonable` and `Arrayable`.

**Comment by X-Coder264** [comment]
I don't understand what you are saying as only the `Arrayable` interface has the `toArray` method :)

**Comment by tomlankhorst** [comment]
Yes, and Jsonable has the `toJson()`  method. So why would we extend the `Paginator`-interface from `Arrayable` and not from `Jsonable`? It seems a bit arbitrary to me.

**Comment by mfn** [comment]
And let's not forget that `\Illuminate\Pagination\Paginator` _already_ `implements Arrayable,…,Jsonable`.

When looking at all the related classes in the namespace `Illuminate\Pagination` I get the impression that the `interface \Illuminate\Contracts\Pagination\Paginator` is **not** meant to decide about the representation of the paginator. I think that's why it has to **no** toArray, toJson, jsonSerializeable.

Only the specific Paginators (not even the AbstractPaginator) implement this.

IMHO this is a deliberate design decision and to me makes sense.

How you represent the paginator is specific to a paginator but not the interface. The interface is concerned about the paginators intrinsics, not how it talks to arbitrary transformation boundaries.

TL;DR: I'm 👎 for this PR.

**Comment by driesvints** [comment]
Like other people have said the Paginator instances already have `Arrayable` implemented. You're here adding methods on a contract which have nothing to do with the contract itself. Contracts should be single purpose which allow you to make sure that concretes implement certain functionality. I fail to see what a Paginator has to do with converting data to an array.

**Comment by driesvints** [comment]
Sorry but we're going to close this one. Thanks for contributing but this one won't happen I'm afraid.

**Comment by tomlankhorst** [comment]
@driesvints `EloquentModel::paginate(..)` returns a `Paginator`-interface that can be used as a request and therefore should be Jsonable/Arrayable. Maybe change the PhpDoc?
```
--- src/Illuminate/Database/Eloquent/Builder.php	(date 1545314501000)
+++ src/Illuminate/Database/Eloquent/Builder.php	(date 1545314501000)
@@ -7,6 +7,7 @@
 use Illuminate\Support\Arr;
 use Illuminate\Support\Str;
 use Illuminate\Pagination\Paginator;
+use Illuminate\Contracts\Support\Jsonable;
 use Illuminate\Contracts\Support\Arrayable;
 use Illuminate\Support\Traits\ForwardsCalls;
 use Illuminate\Database\Concerns\BuildsQueries;
@@ -732,7 +733,7 @@
      * @param  array  $columns
      * @param  string  $pageName
      * @param  int|null  $page
-     * @return \Illuminate\Contracts\Pagination\LengthAwarePaginator
+     * @return \Illuminate\Contracts\Pagination\LengthAwarePaginator|Jsonable|Arrayable
      *
      * @throws \InvalidArgumentException
      */
```

**Comment by driesvints** [comment]
@tomlankhorst actually, I see that the inner `paginator` method returns a `LengthAwarePaginator` concrete so it might be best to change the return type to the concrete class. Same for the `simplePaginate` method.
